stout/http.h

```diff
@@ -0,0 +1,580 @@
+  ) {
+    EventLoop::Buffer& buffer = userdata->buffer;
+
+    buffer += std::string(ptr, size * nmemb);
+
+    return userdata->should_interrupt
+               .get()
+               .load(std::memory_order_relaxed)
+        ? nmemb * size - 1
+        : nmemb * size;
+  }
+
+  friend auto HttpCall(
+      EventLoop&,
+      const std::string&,
+      const Method&,
+      const std::chrono::nanoseconds&,
+      const std::string&);
+};
+
+////////////////////////////////////////////////////////////////////////
+
+class MultiHandle {
+ public:
+  MultiHandle() {
+    CHECK_NOTNULL(multi_ = curl_multi_init());
+  }
+
+  MultiHandle(const MultiHandle& that) = delete;
+  MultiHandle(MultiHandle&& that) {
+    multi_ = that.multi_;
+    that.multi_ = nullptr;
+  }
+
+  ~MultiHandle() {
+    if (multi_ == nullptr) {
+      return;
+    }
+
+    CHECK_EQ(curl_multi_cleanup(multi_), CURLM_OK);
+  }
+
+  CURLM* handle() {
+    CHECK_NOTNULL(multi_);
+    return multi_;
+  }
+
+ private:
+  CURLM* multi_ = nullptr;
+};
+
+////////////////////////////////////////////////////////////////////////
+
+inline auto HttpCall(
+    EventLoop& loop,
+    const std::string& url,
+    const Method& method,
+    const std::chrono::nanoseconds& timeout = std::chrono::nanoseconds(0),
+    const std::string& fields_string = "" // For POST method.
+) {
+  struct Data {
+    EventLoop& loop;
+    std::string url;
+    Method method;
+    std::chrono::nanoseconds timeout;
+    std::string fields_string;
+
+    EventLoop::Timer timer_handle;
+
+    void* k;
+
+    // NOTE: There may be more than one socket created by CURL and each one
+    // should have its own poller.
+    std::unordered_map<curl_socket_t, EventLoop::Poll> curl_context_map;
+
+    EasyHandle easy;
+    MultiHandle multi;
+    EasyHandle::WriteCallbackData context;
+  };
+
+  return loop.Schedule(
+      "http::Get",
+      Eventual<Result>()
+          .context(Data{loop, url, method, timeout, fields_string, loop})
+          .start([](auto& data, auto& k) {
+            using K = std::decay_t<decltype(k)>;
+            data.k = &k;
+
+            uv_handle_set_data(data.timer_handle.base_handle(), &data);
+
+            auto socket_function = +[](
+                                        CURL* easy,
+                                        curl_socket_t sockfd,
+                                        int what,
+                                        void* userp,
+                                        void* socketp) {
+              Data& data = *static_cast<Data*>(userp);
+              auto iterator = data.curl_context_map.find(sockfd);
+              int events = 0;
+
+              auto poll_callback = [](
+                                       uv_poll_t* handle,
+                                       int status,
+                                       int events) {
+                Data& data = *static_cast<Data*>(handle->data);
```

Comment:
I find it hard to read all these lambdas inside lambdas inside lambdas inside lambdas. Could we make these all named functions elsewhere in the file?